### ISSUE TITLE
_.date(new Date( ), "/")  => // Beyond Moment.js?  

### ISSUE THREAD

**Issue by flavioespinoza** [lodash/lodash#4226]
@jdalton First off thanks, and credit where it is due.

I want to give my deepest heartfelt thanks to you, your core development team, and the many contributors who have made lodash my single most utilized and loved utility library. Years ago lodash saved my ass when, as a **_noob-dev_**, I couldn't get my head around "for" and "while"  and  "wtf" loops! :) Lodash made understanding iteration of arrays easy, and I always thought **_"This is how Javascript should have been written."_**  Respect, and mad props!   

I address this specifically to you @jdalton, because I did do my research into closed issues, and this is my first time submitting an enhancement request, so I apologize if I am beating a dead horse, but I am wondering if the addition of a _.date(date, format) is something that perhaps we could have a deeper discussion on?  

I know in issues #1814 and #2714 you deferred [additional date functions to the awesome moment.js](https://github.com/lodash/lodash/issues/2714#issuecomment-252264899), however, a thought occurred to me while struggling with simple date formatting,

> **_"If there were to be something even more awesome than moment.js, would it not be a date method with simple formatting from lodash?"_**

I've used moment.js extensively in the past, but even to this day have I found it overly complex and difficult to use.  Furthermore, having simple Date string conversions say from "03/06/2019" to "Mar 06 2019" so D3 can parse my data array into a timeseries chart is much more difficult than it needs to be. This is true for moment.js and D3's date module.

### In Concept - lodash's date method could look something like this:
```javascript
/**
 *
 * Computes input `date` converts to string and returns with specified `format`.
 *
 * @static
 * @memberOf _
 * @since 5.0.0
 * @category Date
 * @param {date} Date `date` options are `new Date()`, `timestamp` or `string` in valid date format. See example below. (Not sure if I put valid date string formats here)
 * @param {format} String `format` options `/`, `-`, `.`, `full`, `MMM DD YYYY`, `england`, `uk`
 * @returns {string} Returns the date as a String in specified format.
 * @example
 * 
 * // Formatting options
 * let any_date = 'Jan 7, 2019'
 * console.log(date(any_date)) // => 01/07/2019
 * console.log(date(any_date, '/')) // => 01/07/2019
 * console.log(date(any_date, '-')) // => 01-07-2019
 * console.log(date(any_date, '.')) // => 01.07.2019
 * console.log(date(any_date, 'full')) // => Mon Jan 07 2019 00:00:00 GMT-0700 (Mountain Standard Time)
 * console.log(date(any_date, 'MMM DD YYYY')) // => Jan 07 2019
 * console.log(date(any_date, 'england')) // => 07 Jan 2019
 * console.log(date(any_date, 'uk')) // => 07 Jan 2019
 * console.log(date(any_date, undefined)) // => 01/07/2019
 * 
 * // Any timestamp(s) 
 * let five_days_ms = 4.32e+8
 * let now_timestamp = _.now() // => 1551875334611
 * let future_timestamp = _.add(now_timestamp, five_days_ms)
 * let past_timestamp = _.subtract(now_timestamp, five_days_ms)
 * console.log(date(now_timestamp)) // => 03/06/2019
 * console.log(date(future_timestamp)) // => 03/11/2019
 * console.log(date(past_timestamp)) // => 03/01/2019
 *
 * // The addition and subtraction of timestamps show above
 * // could also be a precursor to date math methods:
 * _.dateAdd(date1, date2) 
 * _.dateSubtract(date1, date2)
 *
 */
```

##  Argument against

- **_Why do we need another date formatting library or method?_**
- **_Moment.js and D3.js do all of this and more_**

### This is why: 

- It's the **_MORE_** that is the issue.  **_They try to do too much._**  
- They give the land, dairy farm, cows, etc... when all I want is the milk.   **_All I want is the ESSENTIAL_**

One could argue against lodash's addition of _.now():
> #### "Why do you need a method to get a timestamp for right now?"  

```javascript

const now = _.now()

```

> #### Why not use this:

```javascript

const now = new Date().getTime()

```

> #### Because it is ultra clear on what _.now() does.  

All I needed to do was glance at the documentation -- one time -- and I have never forgotten what _.now() does.

# Why lodash?

> #### Because, Lodash is a Crucible. In it we burn away irrelevancies until we are left with a pure product - the essential, for all time.  

---

### How lodash?

To conclude this little soap-box-opera I have, in fact, tested the **_Concept_** I outlined earlier.  Here it is:

```javascript

const _ = require('lodash')
/**
 *
 * Computes input `date` converts to string and returns with specified `format`.
 *
 * @static
 * @memberOf _
 * @since 5.0.0
 * @category Date
 * @param {date} Date `date` options are `new Date()`, `timestamp` or `string` in valid date format. See example below. (Not sure if I put valid date string formats here)
 * @param {format} String `format` options `/`, `-`, `.`, `full`, `MMM DD YYYY`, `england`, `uk`
 * @returns {string} Returns the date as a String in specified format.
 * @example
 *
 * // If no value is given for both `date` and `format` defaults to now date (example was done on 03/06/19)
 * console.log(date()) // => 03/06/2019
 *
 * // If no value is given for `format` defaults to now date formatted with the `/` option
 * console.log(date('Jan 7, 2019')) // => 01/07/2019
 *
 * // Formatting options
 * let any_date = 'Jan 7, 2019'
 * console.log(date(any_date)) // => 01/07/2019
 * console.log(date(any_date, '/')) // => 01/07/2019
 * console.log(date(any_date, '-')) // => 01-07-2019
 * console.log(date(any_date, '.')) // => 01.07.2019
 * console.log(date(any_date, 'full')) // => Mon Jan 07 2019 00:00:00 GMT-0700 (Mountain Standard Time)
 * console.log(date(any_date, 'MMM DD YYYY')) // => Jan 07 2019
 * console.log(date(any_date, 'england')) // => 07 Jan 2019
 * console.log(date(any_date, 'uk')) // => 07 Jan 2019
 * console.log(date(any_date, 'foo')) // => 01/07/2019
 *
 * // Works with these valid date formats (example was done on 03/06/19)
 * console.log(date()) // => 03/06/2019
 * console.log(date(new Date())) // => 03/06/2019
 * console.log(date(1551875334611)) // => 03/06/2019
 * console.log(date(_.toString(new Date()))) // => 03/06/2019
 * console.log(date('3/6/19')) // => 03/06/2019
 * console.log(date('3-6-19')) // => 03/06/2019
 * console.log(date('3.6.19')) // => 03/06/2019
 * console.log(date('Mar 6 19')) // => 03/06/2019
 * console.log(date('Mar 6, 19')) // => 03/06/2019
 * console.log(date('Mar 6 2019')) // => 03/06/2019
 * console.log(date('Mar 6, 2019')) // => 03/06/2019
 * console.log(date('March 6 19')) // => 03/06/2019
 * console.log(date('March 6, 2019')) // => 03/06/2019
 * console.log(date('6 Mar 2019')) // => 03/06/2019
 * console.log(date('6 March 2019')) // => 03/06/2019
 *
 * // Any date past, present or future
 * let past_date = 'Feb 14, 2019'
 * console.log(date(past_date)) // => 02/14/2019
 * console.log(date(past_date, '/')) // => 02/14/2019
 * console.log(date(past_date, '-')) // => 02-14-2019
 * console.log(date(past_date, '.')) // => 02.14.2019
 * console.log(date(past_date, 'full')) // => Thu Feb 14 2019 00:00:00 GMT-0700 (Mountain Standard Time)
 * console.log(date(past_date, 'MMM DD YYYY')) // => Feb 14 2019
 * console.log(date(past_date, 'england')) // => 14 Feb 2019
 * console.log(date(past_date, 'uk')) // => 14 Feb 2019
 *
 * // Any timestamp
 * let five_days_ms = 4.32e+8
 * let now_timestamp = _.now() // => 1551875334611
 * let future_timestamp = _.add(now_timestamp, five_days_ms)
 * let past_timestamp = _.subtract(now_timestamp, five_days_ms)
 * console.log(date(now_timestamp)) // => 03/06/2019
 * console.log(date(future_timestamp)) // => 03/11/2019
 * console.log(date(past_timestamp)) // => 03/01/2019
 *
 * // Handles undefined or other for 2nd arg 'format'
 * console.log(date(my_date, undefined)) // => 02/14/2019
 * console.log(date(my_date, false)) // => 02/14/2019
 * console.log(date(my_date, true)) // => 02/14/2019
 * console.log(date(my_date, 'foo')) // => 02/14/2019
 * console.log(date(my_date, 8)) // => 02/14/2019
 * console.log(date(my_date, null)) // => 02/14/2019
 */
function date(date, format) {
  let d

  if (!date) {
    d = new Date(_.now())
  } else {
    d = new Date(date)
  }

  const pad = (s) => {
    return s < 10 ? '0' + s : s
  }

  const months = ['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

  let year = d.getFullYear()

  let day = [pad(d.getDate())]

  let month_name = months[d.getMonth() + 1]

  let month_number = pad(d.getMonth() + 1)

  let date_usa = [month_number, day, year]

  let date_uk = `${day} ${month_name} ${year}`

  if (format === '/' || format === '-' || format === '.') {
    return date_usa.join(format)
  } else if (format === 'MMM DD YYYY') {
    return [month_name, day, year].join(' ')
  } else if (format === 'full') {
    return d
  } else if (format === 'england' || format === 'uk') {
    return date_uk
  } else {
    return date_usa.join('/')
  }

}

```

## In summary

Dates are of paramount importance in any coding language, methodology and/or philosophy.

> #### How do we render, manipulate, and represent time?  

Dates require our best -- weather you're on the front-end, backend, or somewhere in-between.  The proficiency required to master Javascript Dates is, perhaps, too fundamental to what we do to leave it to others who, although awesome, haven't sought the perfection that comes, not when nothing more can be added, but when nothing more can be taken away.

### Idea 

Perhaps we could open it to a proposed enhancement with votes needed?  If not, I understand, and I thank you for your time and patients.  

Best,
Flavio

**Comment by jdalton** [comment]
Hi @flavioespinoza!

Thank you for this detailed write-up. I'm still in favor of deferring to more specialized date libraries. However, I think what you have here is great and would make good standalone npm package!
